keps/sig-scheduling/5471-enable-sla-based-scheduling/README.md

```diff
@@ -0,0 +1,1430 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Extend **Node/Device Toleration** to support **numeric comparison operators** when matching **Node/Device Taints**:
+
+**Toleration Enhancements:**
+
+- New operators: `Lt`, `Gt` (in addition to existing `Equal`/`Exists`)
+- Automatic type detection: integer comparison for numeric values, semantic version comparison for semver values
+- Primary motivation: allow pods to opt‑in to nodes by `SLA/failure‑probability` values or version requirements
+
+**CEL Enhancements:**
+
+- Extended semver library with additional comparison and utility functions
+- Enhanced integration with Kubernetes version parsing standards
+- Consistent semantic version handling across validation policies
+
+This creates a unified approach to threshold-based and version-based placement/validation across Kubernetes, preserving the well‑understood safety model of taints/tolerations while enabling sophisticated policy enforcement through CEL expressions.
+
+## Motivation
+
+Many clusters blend (**on‑demand/higher‑SLA**) and (**spot-preemptible/lower‑SLA**) nodes. Platform teams want a safe default keeping most workloads away from risky capacity, while allowing specific workloads to opt‑in with explicit thresholds like `SLA ≥ 95%`.
+
+### Why not NodeAffinity alone?
+
+For the “node SLA / failure‑probability” use‑case, NodeAffinity can express minimum or exact SLA thresholds via label comparisons, but it’s not sufficient for the operational goals here:
+
+- **Policy orientation:** NodeAffinity is per‑pod; to keep most pods away from low‑SLA nodes you'd have to edit every workload.
+- **Taints invert control**: nodes declare risk; only pods with a matching toleration may land.
+- **Eviction semantics:** Affinity has no eviction. Taints support `NoExecute` with `tolerationSeconds`, letting operators drain/evict pods when a node's SLA class drops or a spot reclaim hits.
+- **Operational ergonomics:** Centralized, node‑side policy is consistent with other safety taints (e.g., disk-pressure, memory-pressure). Teams opt‑in, reducing config drift.
+
+From a scheduling perspective, adding numeric operators to tolerations only adjusts match logic. It does not change queueing, scoring, or preemption algorithms.
+
+### Goals
+
+- Add comparison operators to tolerations so pods can match taints like `node.kubernetes.io/sla=<int>` using thresholds.
+- Enable automatic type detection for integer and semantic version comparisons in tolerations.
+- Enhance CEL semver library with additional comparison functions for consistent version handling.
+- Keep behavior consistent with existing effects (`NoSchedule`, `PreferNoSchedule`, `NoExecute`).
+- Provide unified semantic version handling across scheduling and admission control.
+- Backward compatible and opt‑in via feature gates.
```

Comment:
Is it fair to add a goal to ensure that the addition of new operators has zero operational effect on existing pod scheduling outcomes using the Equality or Exists operators? I'm thinking of the various `switch toleration.Operator` bits in source and assuming that adding more operators should have zero cost to existing scenarios.

Reply:
Yes, totally fair. Added a goal in addition to performance tests.